experiment/contractNipopow.sol

```diff
@@ -227,40 +228,56 @@ contract Crosschain {
   }
 
   function compare_proofs(Nipopow storage nipopow,
-    bytes32[] contesting_proof) internal returns(bool) {
+    bytes32[] memory contesting_proof) internal returns(bool) {
     if (nipopow.best_proof.length == 0) {
       return true;
     }
     uint proof_lca_index;
     uint contesting_lca_index;
-    (proof_lca_index, contesting_lca_index) 
+    (proof_lca_index, contesting_lca_index)
       = get_lca(nipopow, contesting_proof);
     return best_arg(nipopow, contesting_proof, contesting_lca_index) >
            best_arg(nipopow, nipopow.best_proof, proof_lca_index);
   }
 
   function verify_merkle(bytes32 roothash, bytes32 leaf, uint8 mu,
-    bytes32[] siblings) pure internal {
+    bytes32[] memory siblings) pure internal {
     bytes32 h = leaf;
     for (uint i = 0; i < siblings.length; i++) {
       uint8 bit = mu & 0x1;
       if (bit == 1) {
-        h = sha256(sha256(siblings[siblings.length-i-1], h));
+        h = sha256(abi.encodePacked(
+                            sha256(abi.encodePacked(siblings[siblings.length-i-1], h)
+                   )));
       } else {
-        h = sha256(sha256(h, siblings[siblings.length-i-1]));
-      } 
+        h = sha256(abi.encodePacked(
+                            sha256(abi.encodePacked(h, siblings[siblings.length-i-1])
+                   )));
+      }
       mu >>= 1;
     }
     require(h == roothash);
   }
 
-  function validate_interlink(bytes32[4][] headers, bytes32[] hashed_headers,
-    bytes32[] siblings) internal pure {
+  // shift bits to the most segnificant byte (256-8 = 248)
+  // and cast it to a 8-bit uint
+  function b32_to_uint8(bytes32 _b) private pure returns (uint8) {
+    return uint8(bytes1(_b << 248));
+  }
+
+  function validate_interlink(bytes32[4][] memory headers,
+    bytes32[] memory hashed_headers,
+    bytes32[] memory siblings) internal pure
+  {
     uint ptr = 0; // Index of the current sibling
     for (uint i = 1; i < headers.length; i++) {
+      // hold the 3rd and 4th least significant bytes
+      bytes32 branch_length_b32 = (headers[i][3] >> 8) & bytes32(uint(0xff));
+      uint8 branch_length = b32_to_uint8(branch_length_b32);
```

Comment:
how would you feel about not extracting the variables here and below?

Reply:
You mean do the elide the call of the function `b32_to_uint8()` or to make these two lines a single one?

Reply:
second